### PR TITLE
Laboratory work #2, Polina Domozhirova - 21FPL2

### PR DESCRIPTION
**Pull request by PolinaDomozhirova**

### CODE REVIEW

File: lab_2_keywords_cooccurrence/main.py [hunk 1]
```diff
@@ -29,7 +45,28 @@ def extract_candidate_keyword_phrases(phrases: Sequence[str], stop_words: Sequen
 
     In case of corrupt input arguments, None is returned
     """
-    pass
+    if not check_list(phrases, str, False) or not check_list(stop_words, str, False):
+        return None
+    candidate_keyword_phrases = []
+    for phrase in phrases:
+        phrase = phrase.lower()
+        phrases1 = phrase.split()
```
Comment (marina-kaz) [inline lab_2_keywords_cooccurrence/main.py:53]: ` phrases1 ` consider naming variables in a more comprehensive way

Comment (marina-kaz) [reply lab_2_keywords_cooccurrence/main.py:53]: i also don't really like names ` first_keywords_candidate` and `second_keywords_candidate`, you can turn it to tuple right before appending to the list

File: lab_2_keywords_cooccurrence/main.py [hunk 2]
```diff
@@ -29,7 +45,28 @@ def extract_candidate_keyword_phrases(phrases: Sequence[str], stop_words: Sequen
...
+        first_keywords_candidate = []
+        for word in phrases1:
+            if word in stop_words:
+                if first_keywords_candidate:
+                    second_keywords_candidate = tuple(first_keywords_candidate)
+                    candidate_keyword_phrases.append(second_keywords_candidate)
+                    first_keywords_candidate.clear()
+            elif word == phrases1[len(phrases1) - 1]:
```
Comment (marina-kaz) [inline lab_2_keywords_cooccurrence/main.py:61]: consider indexing it as `[-1]`

File: lab_2_keywords_cooccurrence/main.py [hunk 3]
```diff
@@ -55,7 +99,17 @@ def calculate_word_degrees(candidate_keyword_phrases: KeyPhrases,
...
     """
-    pass
+    if not check_list(candidate_keyword_phrases, tuple, False) or not  check_list(content_words, str, False):
+        return None
+    dict_degree = {}
+    for phrase in candidate_keyword_phrases:
+        for word in content_words:
+            if word in phrase:
+                dict_degree[word] = len(phrase) + dict_degree.get(word, 0)
+            if word not in dict_degree:
```
Comment (marina-kaz) [inline lab_2_keywords_cooccurrence/main.py:109]: can we avoing checking this condition?

**Review by marina-kaz — COMMENTED** [review]
Good job!
Keep going

File: lab_2_keywords_cooccurrence/main.py (hunk 3 above)
Comment (PolinaDomozhirova) [reply lab_2_keywords_cooccurrence/main.py:109]: do you mean "elif" instead of it?

File: lab_2_keywords_cooccurrence/main.py (hunk 2 above)
Comment (PolinaDomozhirova) [reply lab_2_keywords_cooccurrence/main.py:61]: i've rewrote this function

File: lab_2_keywords_cooccurrence/main.py (hunk 3 above)
Comment (marina-kaz) [reply lab_2_keywords_cooccurrence/main.py]: not exactly 
it is just that this way you are adding this key to the dictionary multiple times 
consider adding them all at once at the very beginning and then just updating

File: lab_2_keywords_cooccurrence/main.py [hunk 4]
```diff
@@ -123,7 +204,28 @@ def extract_candidate_keyword_phrases_with_adjoining(candidate_keyword_phrases:
...
+    frequencies = {}
+    for item in list_of_pairs:
+        frequencies[item] = frequencies.get(item, 0) + 1
+    stop_words_list = []
+    for key_word in frequencies:
+        if frequencies[key_word] > 2:
+            continue
+        for phrase in phrases:
+            if (key_word[0] and key_word[1]) in phrase:
+                stop_words_list.extend(re.findall(f'{key_word[0]}.*{key_word[1]}', phrase))
```
Comment (marina-kaz) [inline lab_2_keywords_cooccurrence/main.py:223]: what if there is more than 1 stop word?

File: lab_2_keywords_cooccurrence/main.py [hunk 5]
```diff
@@ -141,7 +243,17 @@ def calculate_cumulative_score_for_candidates_with_stop_words(candidate_keyword_
...
-    pass
+    if not check_list(candidate_keyword_phrases, tuple, False) or not check_list(stop_words, str, False) \
+            or not check_dict(word_scores, str, Union[int, float], False):
+        return None
+    candidates_cumulative_score = {}
+    for phrase in candidate_keyword_phrases:
+        candidates_cumulative_score[phrase] = 0
+        for word in phrase:
+            if word not in stop_words:
+                candidates_cumulative_score[phrase] += int(word_scores[word])
```
Comment (marina-kaz) [inline lab_2_keywords_cooccurrence/main.py:254]: `int(word_scores[word])` can you explaint turning it to int?

**Review by marina-kaz — COMMENTED** [review]
Good job! 
Some places need clarification though

File: lab_2_keywords_cooccurrence/main.py [hunk 6]
```diff
@@ -84,7 +143,18 @@ def calculate_cumulative_score_for_candidates(candidate_keyword_phrases: KeyPhra
...
-    pass
+    if not check_list(candidate_keyword_phrases, tuple, False) or not  check_dict(word_scores, str, float, False):
+        return None
+    cumulative_score_dict = {}
+    for phrase in candidate_keyword_phrases:
+        cumulative_score = 0
+        for word in phrase:
+            if word not in word_scores:
+                return None
+            cumulative_score += int(word_scores[word])
```
[the remark this answers is not in the document]
Comment (PolinaDomozhirova) [reply lab_2_keywords_cooccurrence/main.py]: i can't explain why i've done it( so it have been fixed.

File: lab_2_keywords_cooccurrence/main.py [hunk 7]
```diff
@@ -123,7 +204,28 @@ def extract_candidate_keyword_phrases_with_adjoining(candidate_keyword_phrases:
...
+        phrase = ' '.join(key_word)
+        joining_phrases.append(phrase)
+    list_of_pairs = list(pairwise(joining_phrases))
+    frequencies = {}
+    for item in list_of_pairs:
+        frequencies[item] = frequencies.get(item, 0) + 1
+    stop_words_list = []
+    for key_word in frequencies:
+        if frequencies[key_word] > 2:
+            continue
```
[the remark this answers is not in the document]
Comment (PolinaDomozhirova) [reply lab_2_keywords_cooccurrence/main.py]: fixed

File: lab_2_keywords_cooccurrence/main.py (hunk 4 above)
Comment (PolinaDomozhirova) [reply lab_2_keywords_cooccurrence/main.py]: fixed

File: lab_2_keywords_cooccurrence/main.py (hunk 5 above)
Comment (PolinaDomozhirova) [reply lab_2_keywords_cooccurrence/main.py]: fixed

**Pull request closed without merging**